### PR TITLE
Add FedJobAPI documentation

### PR DESCRIPTION
**Pull request by SYangster**
- Add FedJobAPI docs
- Clarify and add some docstrings
- Checks for target type for Controller and Executor 
- Add misc 2.5.0 updates

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Quick tests passed locally by running `./runtest.sh`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated.

### CODE REVIEW

File: docs/getting_started.rst [hunk 1]
```diff
@@ -175,6 +175,10 @@ Using any text editor to edit the Dockerfile and paste the following:
 .. literalinclude:: resources/Dockerfile
     :language: dockerfile
 
+.. note::
+
+    For nvflare version 2.3 set PYTORCH_IMAGE=nvcr.io/nvidia/pytorch:23.02-py3
+
```
Comment (YuanTingHsieh) [inline docs/getting_started.rst:181]: Do we need this? Do we want to add this to all the future versions?

File: docs/programming_guide/execution_api_type.rst [hunk 2]
```diff
@@ -35,7 +35,7 @@ The :ref:`client_api` provides the most straightforward way to write FL code,
 and can easily be used to convert centralized code with minimal code changes.
 The Client API uses the :class:`FLModel<nvflare.app_common.abstract.fl_model.FLModel>`
 object for data transfer and supports common tasks such as train, validate, and submit_model.
-Additionally, options for using decorators or PyTorch Lightning are also available.
+Options for in-process and sub-process client api executors are available for different use cases.
```
Comment (YuanTingHsieh) [inline docs/programming_guide/execution_api_type.rst:38]: One track is that, we have:
- client_api (`send/receive`)
- decorator (`@flare.train`)
- PyTorch Lightning (`@flare.patch`)

The other track is whether users want in-process or external process executor.
These are different aspects though

File: docs/programming_guide/execution_api_type/client_api.rst [hunk 3]
```diff
@@ -167,20 +167,26 @@ Client API communication patterns
...
+Communication between them occurs through an efficient in-memory databus.
+
+When the training process involves either a single GPU or no GPUs, and the training script doesn't integrate third-party
+training systems, the in-process executor is preferable (when available).
+
+Sub-process Client API
+----------------------
+
+On the other hand, the LauncherExecutor employs a sub-process to execute training script, leading to the client executor
+and training script residing in separate processes. The "launch_once" option is provided to the SubprocessLauncher to control
```
Comment (YuanTingHsieh) [inline docs/programming_guide/execution_api_type/client_api.rst:184]: A note here the `LauncherExecutor` itself is not using subprocess, it is using the `Launcher` abstraction.

The `SubprocessLauncher` is using subprocess to launch external process.

Some of these abstract/classes we might consolidate with Yan's new PR.

File: tests/unit_test/job_config/fed_job_test.py [hunk 4]
```diff
@@ -15,19 +15,22 @@
...
+        executor = ModelLearnerExecutor(learner_id=job.as_id(ModelLearner()))
 
-        job.to(component, "server")
-        job.to(component, "site-1")
+        job.to(controller, "server")
+        job.to(executor, "site-1")
 
         with pytest.raises(Exception):
-            job.to(component, "site-/1", gpu=0)
+            job.to(executor, "site-/1", gpu=0)
```
Comment (YuanTingHsieh) [inline tests/unit_test/job_config/fed_job_test.py:33]: why this raise exception?

File: nvflare/job_config/fed_job.py [hunk 5]
```diff
@@ -153,10 +159,19 @@ def to(
         self._validate_target(target)
 
         if isinstance(obj, Controller):
-            if target not in self._deploy_map:
-                self._deploy_map[target] = ControllerApp(key_metric=self.key_metric)
-            self._deploy_map[target].add_controller(obj, id)
+            if target != "server":  # add client-side controllers as components
+                if target not in self._deploy_map:
+                    raise ValueError(f"{target} doesn't have an `Executor`. Deploy one first before adding components!")
+                self._deploy_map[target].add_component(obj, id)
```
Comment (YuanTingHsieh) [inline nvflare/job_config/fed_job.py:165]: suggest enhance the error message to:  `raise ValueError(f"{target} doesn't have an `Executor`. Deploy one first before adding client-side controllers!")`
to be more precise.

So users need to first deploy the Executor before they add the Client side controller?

File: docs/programming_guide/fed_job_api.rst [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+Example:
+
+.. code-block:: python
+
+  job.to(Net(), "server")
+
+For unsupported models, the model and persistor can be added as components.
+
+
+Components
```
Comment (YuanTingHsieh) [inline docs/programming_guide/fed_job_api.rst:198]: do we need to talk about client side controller either here or in Controller section?

**Review by YuanTingHsieh — COMMENTED** [review]
Thanks for updating the doc, added some comments.

File: docs/getting_started.rst (hunk 1 above)
Comment (nvkevlu) [reply docs/getting_started.rst:181]: Unfortunately, for grpc version compatibility we do need to provide some way to let users know what is supported, so an alternative would be some sort of table that lists what is compatible

File: docs/programming_guide/fed_job_api.rst (hunk 6 above)
Comment (SYangster) [reply docs/programming_guide/fed_job_api.rst:198]: good idea I will add a section about this

File: tests/unit_test/job_config/fed_job_test.py (hunk 4 above)
Comment (SYangster) [reply tests/unit_test/job_config/fed_job_test.py:33]: "/" is an invalid character for target name

File: docs/getting_started.rst (hunk 1 above)
Comment (SYangster) [reply docs/getting_started.rst:181]: For now it is just these two versions, but in the future it's a good idea to add a table

**Comment by YuanTingHsieh** [comment]
@SYangster can you resolve conflicts? there will be another conflict coming up with PR https://github.com/NVIDIA/NVFlare/pull/2704 as well

**Pull request merged**